### ISSUE TITLE
AMG for large scale heterogeneous FEM problems

### ISSUE THREAD

**Issue by RezaNajian** [ddemidov/amgcl#170]
Dear Denis,

I have had a good experience with AMGCL solver in KratosFramework for fluid and structural problems. 
Recently I have started testing AMG for heterogeneous problems where there are large jumps in material properties within the domain. 

I have been trying different settings that I used before but so far no success in convergence. I have tried combinations of smoothers, krylov solvers, coarsening types. 

Can you give me some hints and tips to use AMG for large scale heterogeneous problems which may suffer from saddle point issue ?

**Comment by ddemidov** [comment]
I think you may need a field-split-type preconditioner for saddle-point problems. See, for example, [1] which was a result of the discussion in #144. It also may be a good idea to try using block-valued backend (`amgcl::backend::builtin<amgcl::static_matrix<double,B,B>>`), or block-wise coarsening (`precond.coarsening.aggr.block_size=B`). Can you share the system matrix and the RHS?

[1] Demidov, Denis, Lin Mu, and Bin Wang. "Accelerating linear solvers for large-scale Stokes problems with C++ metaprogramming." arXiv preprint [arXiv:2006.06052](https://arxiv.org/abs/2006.06052) (2020).

**Comment by rezanajianasl** [comment]
Thanks a lot for the fast response. I have tried AMGCL for two problems: 1. heterogeneous **with** KKT conditions 2.  heterogeneous **without** saddle point.
In both cases I have played and used precond.coarsening.aggr.block_size=B and also smoothers and so on. I must mention that nodal displacements are the only DOFs and KKT conditions are applied with penalty method.

Thanks for sharing the reference. I will share the matrices and RHS as soon as possible.
